File: easy/binary_search.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
 // if 1 value only
 var search = function(nums, target) {
  if (target < nums[0] || target > nums[nums.length-1]) {
      return -1
  }
  
  let [left, right] = [0, nums.length-1]

  while (left <= right) {
      let midway = Math.floor((right + left) / 2)
      if (nums[midway] === target) {
          return midway
      } else if (target < nums[midway]) {
          right = midway - 1
      } else {
          left = midway + 1
      }
  }
  return -1
};
// space O(1)
// time O(logn) (must half the array of n elements logn time)

File: easy/binary_search.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        basically implement binary search

        get middle number index
        is that number, if greater take array lower, if lower take subarray higher
        recursion?
        try using two pointers and a for loop instead, while low >= high
        """

        low = 0
        high = len(nums) - 1
        midway = len(nums) // 2
        
        while low <= high:
            if nums[midway] == target:
                return midway
            elif nums[midway] > target:
                high = midway - 1
                midway = (high + low) // 2
            else: # nums[midway] < target:
                low = midway + 1
                midway = (high + low) // 2
        
        return -1
        # time O(nlogn)
        # space O(1)